### PR TITLE
implement Align::restrict_for_offset and related fn

### PR DESCRIPTION
**Pull request by memoryleak47**

### CODE REVIEW

File: prelude/bigint.md [hunk 1]
```diff
@@ -16,6 +19,9 @@ impl BigInt {
     /// Checks whether `self` is a power of two.
     pub fn is_power_of_two(&self) -> bool;
 
+    /// Returns the absolute value of `self`.
+    pub fn abs(self) -> BigInt;
+
```
Comment (memoryleak47) [inline prelude/bigint.md:24]: whoops, forgot to mention a few BigInt methods..

File: prelude/align.md [hunk 2]
```diff
@@ -23,5 +23,26 @@ impl Align {
     pub fn bytes(self) -> BigInt {
         self.raw
     }
+
+    /// Computes the best alignment possible for the given offset
+    /// (the largest power of two that the offset is a multiple of).
+    ///
+    /// N.B., for an offset of `0`, this returns `2^64`.
```
Comment (memoryleak47) [inline prelude/align.md:30]: Do you prefer another handling of the `offset == zero` case?
This is how rustc does it, but `2^64` is somewhat arbitrary for us - as we are working with BigInt.

Comment (RalfJung) [reply prelude/align.md:30]: This should probably return `Option<Align>`, and return `None` if the offset is zero.

File: prelude/align.md [hunk 3]
```diff
@@ -23,5 +23,26 @@ impl Align {
...
+            None => BigInt::from(64),
+        };
+
+        Align::from_bytes(BigInt::one() << exp)
+    }
+
+    /// Lower the alignment, if necessary, such that the given offset
+    /// is aligned to it (the offset is a multiple of the alignment).
+    pub fn restrict_for_offset(self, offset: Size) -> Align {
+        self.min(Align::max_for_offset(offset))
```
Comment (RalfJung) [inline prelude/align.md:45]: Where is `min` coming from?

File: prelude/bigint.md [hunk 4]
```diff
@@ -7,6 +7,9 @@ Additionally, BigInt provides a few utility functions.
 pub use specr::BigInt;
 
 impl BigInt {
+    pub fn zero() -> BigInt;
+    pub fn one() -> BigInt;
```
Comment (RalfJung) [inline prelude/bigint.md:11]: This overlaps with the other PR, doesn't it?

File: prelude/bigint.md [hunk 5]
```diff
@@ -30,7 +36,7 @@ impl BigInt {
         }
 
         // the modulus.
-        let m = BigInt::from(2).pow(size.bits());
+        let m = BigInt::one() << size.bits();
```
Comment (RalfJung) [inline prelude/bigint.md:39]: Please don't use shifts to encode exponentiation. That's a low-level bit manipulation trick, nothing we want to use in a specification.

File: prelude/align.md (hunk 3 above)
Comment (memoryleak47) [reply prelude/align.md:45]: Align impls `Ord`, I've added a comment for it.

**Comment by memoryleak47** [comment]
@RalfJung ready for review

File: prelude/align.md [hunk 6]
```diff
@@ -3,6 +3,8 @@ See [Align](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_target/abi/str
 
 `Align` is always a power of two.
 
+For convenience, Align implements `Ord`.
```
Comment (RalfJung) [inline prelude/align.md:6]: Yeah but where does that come from? I would have expected a `derive` or so?

**Pull request merged**